### PR TITLE
tg3: fix broadcom NIC 57766 staying down issue

### PR DESCRIPTION
**Pull request by byu343**
This is the workaround for incorrectly detected DMA overflow that may result in NIC staying down after reset. Set consistent dma mask to 31 in the tg3 driver for broadcom NIC ASIC_REV_57766. Before the change, when a previous DMA end with lower 16 bits 0xffff, and a new dma starts with upper bits 0xffff, the hardware seems to think we're at address 0xffffffff and triggers an overflow. The change asks the kernel to only address 31 bits for coherent allocations. This will ensure that upper bits are at most 0x7fff thus avoiding the bug.

This change has been tested on Arista 7060 and 7260 platforms which have the 57766 NIC.

### CODE REVIEW

**Comment by saiarcot895** [comment]
```
2024-08-01T18:32:48.4308213Z /__w/1/s/linux-6.1.38/drivers/net/ethernet/broadcom/tg3.c: In function 'tg3_init_one':
2024-08-01T18:32:48.4340740Z /__w/1/s/linux-6.1.38/drivers/net/ethernet/broadcom/tg3.c:17758:31: error: implicit declaration of function 'pci_set_consistent_dma_mask' [-Werror=implicit-function-declaration]
2024-08-01T18:32:48.4341139Z 17758 |                         err = pci_set_consistent_dma_mask(pdev,
2024-08-01T18:32:48.4341357Z       |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~
2024-08-01T18:32:48.8755815Z   /__w/1/s/linux-6.1.38/scripts/check-local-export drivers/gpu/drm/drm_gem_vram_helper.o
```

**Comment by prgeor** [comment]
@byu343 please check the build failure

**Review by paulmenzel — CHANGES_REQUESTED** [review]
> Add kernel patch to fix 57766 staying down after reset

What does *57766* mean? Please improve the description.

File: patch/driver-net-tg3-change-dma-mask-for-57766.patch [hunk 1]
```diff
@@ -0,0 +1,30 @@
+diff --git a/drivers/net/ethernet/broadcom/tg3.c b/drivers/net/ethernet/broadcom/tg3.c
```
[the remark this answers is not in the document]
Comment (byu343) [reply patch/driver-net-tg3-change-dma-mask-for-57766.patch:19]: Done

**Comment by byu343** [comment]
> ```
> 2024-08-01T18:32:48.4308213Z /__w/1/s/linux-6.1.38/drivers/net/ethernet/broadcom/tg3.c: In function 'tg3_init_one':
> 2024-08-01T18:32:48.4340740Z /__w/1/s/linux-6.1.38/drivers/net/ethernet/broadcom/tg3.c:17758:31: error: implicit declaration of function 'pci_set_consistent_dma_mask' [-Werror=implicit-function-declaration]
> 2024-08-01T18:32:48.4341139Z 17758 |                         err = pci_set_consistent_dma_mask(pdev,
> 2024-08-01T18:32:48.4341357Z       |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~
> 2024-08-01T18:32:48.8755815Z   /__w/1/s/linux-6.1.38/scripts/check-local-export drivers/gpu/drm/drm_gem_vram_helper.o
> ```

This has been fixed. It was for some kernel API change on bookworm.

**Review by vvolam — COMMENTED** [review]
@byu343 Is the testing completed? Could you please update test information in the PR description?

File: patch/driver-net-tg3-change-dma-mask-for-57766.patch [hunk 2]
```diff
@@ -0,0 +1,51 @@
+From 821f6d79ad2773e0ff1537c0bb3c7af93a694709 Mon Sep 17 00:00:00 2001
+From: Boyang Yu <byu@arista.com>
+Date: Fri, 9 Aug 2024 17:03:51 +0000
+Subject: [PATCH] [PATCH] tg3: fix broadcom NIC 57766 staying down issue
```
[the remark this answers is not in the document]
Comment (byu343) [reply patch/driver-net-tg3-change-dma-mask-for-57766.patch]: Yes, the test has been completed and there is no regression. Duplicate of [PATCH] is fixed.

File: patch/driver-net-tg3-change-dma-mask-for-57766.patch [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ 				"No usable DMA configuration, aborting\n");
+ 			goto err_out_apeunmap;
+ 		}
++
++		if (tg3_asic_rev(tp) == ASIC_REV_57766) {
++			err = dma_set_coherent_mask(&pdev->dev,
++						    DMA_BIT_MASK(31));
++			if (err < 0) {
++				dev_err(&pdev->dev,
++					"Unable to obtain 31 bit DMA for consistent allocations\n");
```
Comment (vvolam) [inline patch/driver-net-tg3-change-dma-mask-for-57766.patch:42]: nit: please check if this line is crossing 120 column width limit.

**Review by vvolam — APPROVED** [review]
A small nit pick, otherwise LGTM. Thank you for the patch.

File: patch/driver-net-tg3-change-dma-mask-for-57766.patch (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulmenzel) [reply patch/driver-net-tg3-change-dma-mask-for-57766.patch:19]: Have you sent this patch upstream?

**Review by prgeor — APPROVED** [review]

File: patch/driver-net-tg3-change-dma-mask-for-57766.patch (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulmenzel) [reply patch/driver-net-tg3-change-dma-mask-for-57766.patch:19]: By the way, thank you for the great commit message.

[the remark this answers is not in the document]
Comment (byu343) [reply patch/driver-net-tg3-change-dma-mask-for-57766.patch:19]: We were told that Broadcom will send it to upstream.

File: patch/driver-net-tg3-change-dma-mask-for-57766.patch (hunk 3 above)
Comment (byu343) [reply patch/driver-net-tg3-change-dma-mask-for-57766.patch:42]: I just checked. It has 99 columns, so is not above 120.

**Review by saiarcot895 — APPROVED** [review]

File: patch/driver-net-tg3-change-dma-mask-for-57766.patch (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulmenzel) [reply patch/driver-net-tg3-change-dma-mask-for-57766.patch:19]: Please verify that, and report back the timeframe. (Vendors like Nokia failed to keep promises in the past unfortunately.)

**Pull request merged**